Sources/SentryCrash/Recording/Tools/SentryCrashCxaThrowSwapper.c

```diff
@@ -0,0 +1,423 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "SentryCrashCxaThrowSwapper.h"
+
+#include <dlfcn.h>
+#include <errno.h>
+#include <execinfo.h>
+#include <mach-o/dyld.h>
+#include <mach-o/nlist.h>
+#include <mach/mach.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/types.h>
+
+#include "SentryAsyncSafeLog.h"
+#include "SentryCrashMach-O.h"
+#include "SentryCrashPlatformSpecificDefines.h"
+
+typedef struct {
+    uintptr_t image_dli_fbase_address;
+    uintptr_t cxa_throw_original_function;
+} SentryCrashImageToOriginalCxaThrowPair;
+
+static cxa_throw_type g_cxa_throw_handler = NULL;
+static const char *const g_cxa_throw_name = "__cxa_throw";
+
+static SentryCrashImageToOriginalCxaThrowPair *g_cxa_originals = NULL;
+static size_t g_cxa_originals_capacity = 0;
+static size_t g_cxa_originals_count = 0;
+
+static uintptr_t
+findOriginalCxaThrowFunction(uintptr_t image_dli_fbase_address)
+{
+    SENTRY_ASYNC_SAFE_LOG_TRACE(
+        "Finding original __cxa_throw for image with base address %p", image_dli_fbase_address);
+
+    for (size_t i = 0; i < g_cxa_originals_count; i++) {
+        if (g_cxa_originals[i].image_dli_fbase_address == image_dli_fbase_address) {
+            return g_cxa_originals[i].cxa_throw_original_function;
+        }
+    }
+    SENTRY_ASYNC_SAFE_LOG_WARN("Address %p not found", image_dli_fbase_address);
+    return (uintptr_t)NULL;
+}
+
+static void
+addPair(SentryCrashImageToOriginalCxaThrowPair pair)
+{
+    uintptr_t originalCxaThrowFunction = findOriginalCxaThrowFunction(pair.image_dli_fbase_address);
+    if (originalCxaThrowFunction != (uintptr_t)NULL) {
+        SENTRY_ASYNC_SAFE_LOG_DEBUG("Already added address pair with image with base address: %p, "
+                                    "and originalCxaThrowFunction: %p",
+            (void *)pair.cxa_throw_original_function, (void *)pair.cxa_throw_original_function);
+        return;
+    }
+
+    SENTRY_ASYNC_SAFE_LOG_DEBUG(
+        "Adding pair for image with base address: %p, and originalCxaThrowFunction: %p",
+        (void *)pair.cxa_throw_original_function, (void *)pair.cxa_throw_original_function);
+
+    if (g_cxa_originals_count == g_cxa_originals_capacity) {
+        g_cxa_originals_capacity *= 2;
+        g_cxa_originals = (SentryCrashImageToOriginalCxaThrowPair *)realloc(g_cxa_originals,
```

Comment:
`m`: IIUC if this realloc returns NULL it has already overwritten the `g_cxa_originals`. Not sure if this has an impact.

Reply:
Good catch. The [Cpp reference on realloc](https://en.cppreference.com/w/c/memory/realloc) states: 

> On failure, returns a null pointer. The original pointer ptr remains valid and may need to be deallocated with [free](https://en.cppreference.com/w/c/memory/free.html) or realloc.

According to my understanding, as we pass `g_cxa_originals` as the pointer and then assign the result to it, we would leak memory if `realloc` fails, because it returns NULL, overwrites `g_cxa_originals` to `NULL` and we loose the pointer to it's previous memory. @supervacuus can you help us out here please?